File: src/functions/login-bounce/loginbounce.go
package loginbounce

import (
	"bytes"
	"net/http"
	"text/template"

	"github.com/segmentio/ksuid"

	"github.com/roleypoly/roleypoly/src/common"
	"github.com/roleypoly/roleypoly/src/common/faas"
)

var (
	redirectPathTemplate = template.Must(
		template.New("redirect").Parse(
			`https://discord.com/api/oauth2/authorize?client_id={{.ClientID}}&scope=identify,guilds&redirect_uri={{.RedirectURI}}&state={{.State}}`,
		),
	)
	clientID    = common.Getenv("BOT_CLIENT_ID").String()
	redirectURI = common.Getenv("OAUTH_REDIRECT_URI").String()
)

type redirectPathData struct {
	ClientID    string
	RedirectURI string
	State       string
}

func LoginBounce(rw http.ResponseWriter, r *http.Request) {
	faas.Stash(rw, r.URL.Query().Get("redirect_url"))

	pathData := redirectPathData{
		ClientID:    clientID,
		RedirectURI: redirectURI,
		State:       ksuid.New().String(),
	}

	pathBuffer := bytes.Buffer{}
	redirectPathTemplate.Execute(&pathBuffer, pathData)

	faas.Bounce(rw, pathBuffer.String())
}

File: src/backend-worker/handlers/login-bounce.ts
import { v4 as uuidv4 } from 'uuid';
import { Bounce } from '../utils/bounce';

type URLParams = {
    clientID: string;
    redirectURI: string;
    state: string;
};

const buildURL = (params: URLParams) =>
    `https://discord.com/api/oauth2/authorize?client_id=${
        params.clientID
    }&response_type=code&scope=identify%20guilds&redirect_uri=${encodeURIComponent(
        params.redirectURI
    )}&state=${params.state}`;

export const LoginBounce = (request: Request): Response => {
    const state = uuidv4();
    const redirectURI = `${API_PUBLIC_URI}/login-callback`;
    const clientID = BOT_CLIENT_ID;

    return Bounce(buildURL({ state, redirectURI, clientID }));
};
